File: blockchain/chain.py
import hashlib
import struct
from time import time
from typing import List

from flask import Flask


class Transaction:
    def __init__(self, orig: str, dest: str, amount: float):
        self.orig = orig
        self.dest = dest
        self.amount = amount

    def hash(self) -> bytes:
        m = hashlib.md5()
        m.update(self.orig.encode('UTF8'))
        m.update(self.dest.encode('UTF8'))
        m.update(struct.pack('!d', self.amount))
        return m.digest()


class Block:
    def __init__(self, transactions: List[Transaction] = None, previous_hash: bytes = b''):
        self.timestamp = time()
        self.transactions = transactions
        self.previous_hash = previous_hash
        self.proof = 0

    def hash(self) -> bytes:
        m = hashlib.md5()
        m.update(struct.pack('!d', self.timestamp))
        if self.transactions is not None:
            for transaction in self.transactions:
                m.update(transaction.hash())
        m.update(self.proof.to_bytes(16, byteorder='big'))
        m.update(self.previous_hash)
        return m.digest()

    def mine(self, length: int = 2):
        array = b'\x00' * length
        while self.hash()[:length] != array:
            self.proof += 1
        print(self.hash())


class Chain:
    def __init__(self, app: Flask = None):
        self.app = None
        self.blocks: List[Block] = [Block()]
        self.transactions: List[Transaction] = []
        if app is not None:
            self.init_app(app)

    def init_app(self, app: Flask):
        self.app = app

    def add_block(self):
        block = Block(self.transactions, previous_hash=self.blocks[-1].hash())
        block.mine()
        self.blocks.append(block)
        self.transactions = []

    def add_transaction(self, orig: str, dest: str, amount: float):
        self.transactions.append(Transaction(orig, dest, amount))


chain = Chain()

if __name__ == '__main__':
    chain = Chain()
    chain.add_transaction("jp", "annie", 10.0)
    chain.add_transaction("annie", "martin", 5.0)
    chain.add_transaction("jp", "martin", 2.0)
    chain.add_block()

File: src/chain.ts
export class Transaction {
    readonly orig: string
    readonly dest: string
    readonly amount: number
    readonly _hash: number

    constructor(orig: string, dest: string, amount: number) {
        this.orig = orig
        this.dest = dest
        this.amount = amount
        this._hash = Math.random()
    }

    hash(): number {
        return this._hash
    }
}

export class Block {
    readonly transactions: Transaction[]
    readonly previousHash: number
    _hash: number

    constructor(transactions: Transaction[] = [], previousHash: number = 0) {
        this.transactions = transactions
        this.previousHash = previousHash
        this._hash = 0
    }

    async mine(maxSeconds: number): Promise<void> {
        return new Promise<void>(resolve => setTimeout(() => {
            this._hash = Math.random()
            resolve()
        }, Math.round(Math.random() * 1000 * maxSeconds)))
    }

    isMined(): boolean {
        return this._hash !== 0
    }

    hash(): number {
        return this._hash
    }
}

class Chain{
    readonly blocks: Block[] = []

    constructor() {
        const initialBlock = new Block()
        initialBlock.mine(5).then(() => this.blocks.push(initialBlock))
    }
}